orca-keel/src/main/kotlin/com/netflix/spinnaker/orca/keel/task/PublishDeliveryConfigTask.kt

```diff
@@ -0,0 +1,157 @@
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      } else if (trigger.branch != null && context.ref == null) {
+        context.ref = "refs/heads/${trigger.branch}"
+        log.debug("Inferred context.ref from trigger: ${context.ref}")
+      }
+      if (context.scmType == null) {
+        context.scmType = trigger.source
+        log.debug("Inferred context.scmType from trigger: ${context.scmType}")
+      }
+      if (context.project == null) {
+        context.project = trigger.project
+        log.debug("Inferred context.project from trigger: ${context.project}")
+      }
+      if (context.repository == null) {
+        context.repository = trigger.slug
+        log.debug("Inferred context.repository from trigger: ${context.repository}")
+      }
+    } else {
+      if (context.ref == null) {
+        context.ref = "refs/heads/master"
+      }
+      if (context.scmType == null || context.project == null || context.repository == null) {
+        throw IllegalArgumentException("scmType, project and repository are required fields in the stage if there's no git trigger.")
+      }
+    }
+
+    val manifestLocation = "${context.scmType}://${context.project}/${context.repository}/.netflix/${context.directory
+      ?: ""}/${context.manifest}@${context.ref}"
+
+    return try {
+      log.info("Retrieving keel manifest at $manifestLocation")
+      val deliveryConfig = scmService.getDeliveryConfigManifest(
+        context.scmType, context.project, context.repository, context.directory, context.manifest, context.ref)
+
+      log.info("Publishing manifest ${context.manifest} to keel on behalf of $user")
+      keelService.publishDeliveryConfig(deliveryConfig, user)
+
+      TaskResult.builder(ExecutionStatus.SUCCEEDED).context(emptyMap<String, Any?>()).build()
+    } catch (e: RetrofitError) {
+      when {
+        e.kind == RetrofitError.Kind.NETWORK -> {
+          // retry if unable to connect
+          log.error("network error talking to downstream service, attempt ${context.attempt} of ${context.maxRetries}: ${e.friendlyMessage}")
+          buildRetry(context)
+        }
+        e.response?.status == HttpStatus.NOT_FOUND.value() -> {
+          // just give up on 404
+          val errorDetails = "404 response from downstream service, giving up: ${e.friendlyMessage}"
+          log.error(errorDetails)
+          buildError(errorDetails)
+        }
+        else -> {
+          // retry on other status codes
+          log.error("HTTP error talking to downstream service, attempt ${context.attempt} of ${context.maxRetries}: ${e.friendlyMessage}")
+          buildRetry(context)
+        }
+      }
+    } catch (e: Exception) {
+      log.error("Unexpected exception while executing {}, aborting.", javaClass.simpleName, e)
+      buildError(e.message)
+    }
+  }
+
+  override fun getBackoffPeriod() = TimeUnit.SECONDS.toMillis(30)
+
+  override fun getTimeout() = TimeUnit.SECONDS.toMillis(180)
```

Comment:
I didn't think these (or `MAX_RETRIES`) needed to be configurable for now, but let me know.